src/ServiceControl/Monitoring/MonitorEndpointsFoundInFailedMessageHeaders.cs

```diff
@@ -0,0 +1,44 @@
+﻿namespace ServiceControl.EndpointControl.Handlers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Monitoring;
+    using NServiceBus.Transport;
+    using Operations;
+    using ServiceControl.Contracts.Operations;
+
+    class MonitorEndpointsFoundInFailedMessageHeaders : IErrorMessageBatchPlugin
+    {
+        public Task AfterProcessing(List<MessageContext> batch)
+        {
+            var observedEndpoints = new HashSet<EndpointDetails>();
+            foreach (var context in batch)
+            {
+                var errorEnricherContext = context.Extensions.Get<ErrorEnricherContext>();
+
+                void RecordKnownEndpoint(string key)
```

Comment:
It's funny with local functions and heap allocations. There is a lot of mad science going and edge cases where it does or does not heap allocate

> Depending on their use, local functions can avoid heap allocations that are always necessary for lambda expressions. If a local function is never converted to a delegate, and none of the variables captured by the local function are captured by other lambdas or local functions that are converted to delegates, the compiler can avoid heap allocations.

I generally tend to avoid them or directly mark them as static

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions#heap-allocations